src/Aspire.Hosting/ApplicationModel/ExpressionResolver.cs

```diff
@@ -128,30 +145,32 @@ await EvalEndpointAsync(endpointReference, EndpointProperty.Port).ConfigureAwait
             }
         }
 
-        return value;
+        return new ResolvedValue(value, false);
     }
 
     /// <summary>
     /// Resolve an expression when it is being used from inside a container.
     /// So it's either a container-to-container or container-to-exe communication.
     /// </summary>
-    async ValueTask<string?> ResolveInternalAsync(object? value)
+    async ValueTask<ResolvedValue> ResolveInternalAsync(object? value)
     {
-        return value switch
+        return (value, sourceIsContainer) switch
         {
-            ConnectionStringReference cs => await ResolveInternalAsync(cs.Resource.ConnectionStringExpression).ConfigureAwait(false),
-            IResourceWithConnectionString cs => await ResolveInternalAsync(cs.ConnectionStringExpression).ConfigureAwait(false),
-            ReferenceExpression ex => await EvalExpressionAsync(ex).ConfigureAwait(false),
-            EndpointReference endpointReference => await EvalEndpointAsync(endpointReference, EndpointProperty.Url).ConfigureAwait(false),
-            EndpointReferenceExpression ep => await EvalEndpointAsync(ep.Endpoint, ep.Property).ConfigureAwait(false),
-            IValueProvider vp => await EvalValueProvider(vp).ConfigureAwait(false),
+            (ConnectionStringReference cs, true) => await ResolveInternalAsync(cs.Resource.ConnectionStringExpression).ConfigureAwait(false),
```

Comment:
I am afraid that after this change the code is not covering all the cases that the previous implementation did. For example, in the previous implementation the `ConnectionStringReference` was matching unconditionally. In the new implementation it only matches for containers.

Reply:
Also see my comment about introducing `sourceIsContainer` parameter above

Reply:
> ConnectionStringReference was matching unconditionally. In the new implementation it only matches for containers

Previously, that code path would also never be run for non-containers. If you see the old version of ExpressionResolver#166 (the only entrypoint into the resolver), non-container cases only resolve the value from the value provider
![image](https://github.com/user-attachments/assets/626b6f80-0ee1-4d7f-b05c-1ca56a51465a)

Reply:
Got it. Thanks for clarification